website_form_builder/README.rst

```diff
@@ -0,0 +1,106 @@
+.. image:: https://img.shields.io/badge/license-LGPL--3-blue.svg
+   :target: https://www.gnu.org/licenses/lgpl
+   :alt: License: LGPL-3
+
+====================
+Website Form Builder
+====================
+
+This module provides websites the feature of adding custom forms in any page.
+
+Configuration
+=============
+
+To configure this module, you need to:
+
+#. Have *Administration / Settings* privileges.
+#. Go to *Settings > Activate developer mode*.
+#. Go to *Settings > Technical > Database Structure > Models*.
+#. Search for the model you want to manage website form access for.
+#. When you find it, it will have a *Website Forms* section where you can:
+
+   * Allow the model to get forms, by checking *Allowed to use in forms*.
+   * Give the model forms a better name in *Label for form action*.
+   * Choose the field where to store custom fields data in *Field for custom
+     form data*. If you leave this one empty and the model is a mail thread,
+     a new message will be appended with that custom data.
+
+#. In the *Fields* tab, there's a new column called *Blacklisted in web forms*.
+   It's a security feature that forbids form submitters to write to those
+   fields. When you create a new website form, all its model fields are
+   automatically whitelisted for the sake of improving the UX. If you want to
+   have higher control, come back here after creating the form and blacklist
+   any fields you want.
```

Comment:
I prefer this approach for not having to manually add each of the fields we wish to add or supply glue modules for each possible model.

Reply:
(moved inline)

Hmmm but modules that are naive of forms, which inherit a model that is be exposed by a form,
typically expect that the data is coming from an authenticated source. An example is prescription or doctor verification. This data can now be submitted by users that are not authenticated.

Reply:
Models are opt-in, fields are opt-out

Reply:
In the example I was mentioning, the validation field is on the partner. I see a lot of potential uses for opting the partner in, which would opt in those fields too.

I'm actually kind of confused though. This black list is circumventing the pre-existing whitelist of `website_form`? Why is the whitelist also being used [in the demo data](https://github.com/OCA/website/pull/402/files#diff-3919e194d50de1418256b036abcbda13R13)?

Reply:
Well, the point here is that, in an awesomely rare exception, `website_form` is completely undocumented (🤣). I tried to reuse the existing engine as much as possible, and it actually makes sense. Let me explain what it does:

- `ir.model` has [some fields to whitelist the model to be used in website forms](https://github.com/odoo/odoo/blob/cfe0ebcc59318a931e0f2592283ac933ee842ba3/addons/website_form/models/models.py#L19-L21).
- [`ir.model.fields` has a `website_form_blacklisted` boolean that defaults to `True`](https://github.com/odoo/odoo/blob/cfe0ebcc59318a931e0f2592283ac933ee842ba3/addons/website_form/models/models.py#L107), so all model fields are blacklisted by default.
- [When a model is not whiltelisted, the controller aborts](https://github.com/odoo/odoo/blob/cfe0ebcc59318a931e0f2592283ac933ee842ba3/addons/website_form/controllers/main.py#L24). No models are whitelisted by default in `website_form` addon.
- [When a field is blacklisted, but the model is whitelisted, the field is considered *extra data* (like a custom field that does not exist)](https://github.com/odoo/odoo/blob/cfe0ebcc59318a931e0f2592283ac933ee842ba3/addons/website_form/controllers/main.py#L134-L144).
- The `formbuilder_whitelist` method is a [public method only available to website designers](https://github.com/odoo/odoo/blob/cfe0ebcc59318a931e0f2592283ac933ee842ba3/addons/website_form/models/models.py#L93), which means that those users should be able to decide which fields they want to make writeable. The most UX-friendly way of expressing such desire is when you add a field to a website form, so [when you save the form, any model fields found there are whitelisted automatically](https://github.com/OCA/website/pull/402/files#diff-95cb4eb71cc38363ac6f31057c94ad32R185).
- [Core addons use such method to enable specific models](https://github.com/odoo/odoo/search?l=XML&q=formbuilder_whitelist&type=&utf8=%E2%9C%93), so when you install those addons you automatically get a website-form-enabled model ready to use out of the box.
- I use such method in demo data to mimic the behavior of such addons and have some data to run the tour (which is ❌ right now).
- The main purpose of `website_form` is to allow unauthenticated users to create records that they should have no access to, such as creating a `crm.lead`. The way to verify that the form came from a trusted source we have the CSRF token, provided out of the box. If we want to use the engine to provide a more granular authentication system, such as allowing only portal users to submit to certain models, then I think that belongs to another addon, independent of this one (it would extend the `website_form` engine, no matter if to use it with EE builder or this one).